cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java

```diff
@@ -0,0 +1,109 @@
+            if (!canRoll) {
+                LOGGER.debugCr(reconciliation, "Restart pod {} would affect the quorum", podId);
+            }
+            return canRoll;
+        }).recover(error -> {
+            LOGGER.warnCr(reconciliation, "Error determining whether it is safe to restart pod {}", podId, error);
+            return Future.failedFuture(error);
+        });
+    }
+
+    /**
+     * Returns id of the quorum leader.
+     **/
+    Future<Integer> quorumLeaderId() {
+        LOGGER.debugCr(reconciliation, "Determining the quorum leader id");
+        if (quorumInfoFuture == null) {
+            this.quorumInfoFuture = describeMetadataQuorum();
+        }
+        return quorumInfoFuture.map(info -> info.leaderId()).recover(error -> {
+            LOGGER.warnCr(reconciliation, "Error determining the quorum leader id", error);
+            return Future.failedFuture(error);
+        });
+    }
+
+    private boolean isQuorumHealthy(int leaderId, int podId,  Map<Integer, Long> replicaStates) {
+        int totalNumOfControllers = replicaStates.size();
+        AtomicInteger numOfCaughtUpControllers = new AtomicInteger();
+        if (leaderId < 0) {
+            LOGGER.warnCr(reconciliation, "No quorum leader is found because the leader id is set to {}", leaderId);
+            return false;
+        }
+        long leaderLastCaughtUpTimestamp = replicaStates.get(leaderId);
+        LOGGER.debugCr(reconciliation, "The lastCaughtUpTimestamp for the leader replica {} is {}", leaderId, leaderLastCaughtUpTimestamp);
+        replicaStates.forEach((replicaId, lastCaughtUpTimestamp) -> {
+            if (lastCaughtUpTimestamp < 0) {
+                LOGGER.warnCr(reconciliation, "No valid lastCaughtUpTimestamp is found for the replica {} ", replicaId);
+            } else {
+                if ((leaderLastCaughtUpTimestamp - lastCaughtUpTimestamp) < controllerQuorumFetchTimeoutMs) {
+                    // skip the controller that we are considering to roll unless it's the leader
+                    if (replicaId != podId || replicaId == leaderId) {
+                        numOfCaughtUpControllers.getAndIncrement();
+                    }
+                    LOGGER.debugCr(reconciliation, "The controller {} has caught up with the quorum leader", replicaId);
+                    LOGGER.debugCr(reconciliation, "The lastCaughtUpTimestamp for controller {} is {}", replicaId, lastCaughtUpTimestamp);
+                } else {
+                    LOGGER.debugCr(reconciliation, "The controller {} is behind from the leader", replicaId);
+                }
+            }
+        });
+        LOGGER.debugCr(reconciliation, "Out of {} voters, there are {} controllers that have caught up with the quorum leader, not including the controller {}", totalNumOfControllers, numOfCaughtUpControllers, podId);
+        return numOfCaughtUpControllers.get() >= ceil((double) (totalNumOfControllers + 1) / 2);
```

Comment:
I'm not too fixated on it ... but would this be more readable?

```suggestion
        return numOfCaughtUpControllers.get() > ceil((double) (totalNumOfControllers) / 2);
```

Reply:
Potentially, although do we then need the ceil function at all? What we have now matches the proposal, but happy to change it

Reply:
Not sure about the `ceil`. I would need unit test it I think :-D

Reply:
> If numOfCaughtUpControllers is 2 and totalNumOfControllers is 3, then ceil((double) (totalNumOfControllers) / 2) is 2 which isn't > numOfCaughtUpControllers.

Maybe I don't follow what this method does. But it is named `isQuorumHealthy`: which suggests checking for quorum right now. But it is used to define the variable `canRoll` and whether the quorum is healthy does not define whether you can roll the next controller or not.

If it is used to indicate healthiness, then sure, 2 caught-up nodes out of 3 means it is healthy. But if it should determine rollability then 2 caught-up controllers out of 3 means that you cannot roll because you could roll one of the caught-up controllers and make the quorum unhealthy. So at least for the second case, this should actually return false in this case.

Reply:
I think it's poorly named, the calculation is ignoring the node it is intending to roll, so really it's `isQuorumHealthyWithoutThisPod`, but I'm just trying to work out why we count the current pod if it's the active controller..

Reply:
I've pushed an update to the function to rename it and only ignore the pod we are trying to roll

Reply:
I think you forgot to push, but LGTM as described.

Reply:
I've already pushed the change to rename the function and only ignore the pod we are trying to roll, I haven't made any change to the actual calculation line